### DISCUSSION TITLE
Installing 2.2.5 on Ubuntu 23.10 Mantic

### DISCUSSION THREAD

**Discussion by Radio-biscuit** [savonet/liquidsoap#3925] [Installation]
Hello everyone,

After leaving my liquidsoap setup for a few months, I'm trying to get it working again and found that I must have removed it for some reason that I can't remember.

So, I'm installing again and want to use 2.2.5. However, I'm using Ubuntu 23.10 and the dependencies show that I am using later versions of some libraries. I've tried to override the dependencies, but being a bit of a noob, I can't get it to install.

Does anyone know of a SIMPLE list of things I should do to get the package back onto my system again? I can't even get the version that works with my setup to install (2.2.1). I may have broken everything.

I'm using the 2.2.5 package for Ubuntu Luna when in fact I have Mantic. I need to get around this, so any help would be really useful.

Thanks in advance for your guidance.

**Comment by vitoyucepi** [comment]
Hi @Radio-biscuit,

1. > I've tried to override the dependencies, but being a bit of a noob, I can't get it to install.

   Sometimes it's good to understand your limitations and hire a professional, especially when developing a commercial product, especially when dealing with people's personal information, especially in the publicly available service.
1. > So, I'm installing again and want to use 2.2.5. However, I'm using Ubuntu 23.10 and the dependencies show that I am using later versions of some libraries.
   > I'm using the 2.2.5 package for Ubuntu Luna when in fact I have Mantic. I need to get around this, so any help would be really useful.

   Sorry, but liquidsoap doesn't provide packages for mantic. Here's an installation guide for reference https://www.liquidsoap.info/doc-2.2.4/install.html.
   I can recommend you to choose from the following options:
   1. Use `ubuntu:22.04`, even it's 2 years old, but it's fine to use it on the production server, even after `ubuntu:24.04` has been released. You can get the packages from [the release artifacts](https://github.com/savonet/liquidsoap/releases).
   1. Use `ubuntu:24.04`, but install `liquidsoap:2.2.4` from the ubuntu repos.
   1. Use containers and images if you're familiar with them. Liquidsoap has official docker container images available at https://hub.docker.com/r/savonet/liquidsoap.
   1. Install from opam. There's a guide for this at https://www.liquidsoap.info/doc-2.2.4/install.html#install-using-opam. I recommend using ocaml 4, because there's a serious increase in memory consumption in ocaml 5.

1. > I'm using Ubuntu 23.10

   I can't recommend using non-LTS Ubuntu releases for production, because their support level is insufficient, unless you actively maintain everything yourself. However, it's fine to use them for your workstation.

1. > Does anyone know of a SIMPLE list of things I should do to get the package back onto my system again? I can't even get the version that works with my setup to install (2.2.1). I may have broken everything.

   It's not that easy, because nobody knows what the current state of your system is, I can help you, but it will take some time to find out what went wrong.

**Comment by Radio-biscuit** [comment]
"Sometimes it's good to understand your limitations" er, yeah, this is a hobby and I have memory problems. Hiring a professional isn't something that I'd do. Be careful when assuming context, as I have worked with professionals that have wrecked systems in the past. Please don't assume anything about the person you are talking to, even if they do say "noob"!

Anyway, thanks for the list of possibilities here. This is exactly what I needed when I was thinking of a "simple" solution. I'm going to go back to 22.04, but I would use 24.04 if there was a current upgrade (that won't be available until at least August I understand).

Seriously, you've really helped and I thank you for that!

**Reply by vitoyucepi** [reply]
> "Sometimes it's good to understand your limitations" er, yeah, this is a hobby and I have memory problems. Hiring a professional isn't something that I'd do. Be careful when assuming context, as I have worked with professionals that have wrecked systems in the past. Please don't assume anything about the person you are talking to, even if they do say "noob"!

No offense, just a suggestion. In the modern world, where any misconfigured instance can become part of a botnet in a matter of days, it's important to warn. Even the professional programmer with 10 years of experience can make certain types of mistakes that port scanner enjoyers will welcome.

**Reply by Radio-biscuit** [reply]
Agreed. Thanks for the precautions!

**Reply by gAlleb** [reply]
Yeap, true. 

@vitoyucepi, hi! I'm learning some more about docker. I now migrated from opam to dcker LS with

```ruby
services:
  liquidsoap:
    image: savonet/liquidsoap:v2.2.5
    user: 1001:1001
    command:
      - /home/radio/liquidsoap/omfm/index.liq
    volumes:
      - ./:/home/radio
      - /var/www/html/omfm/hls:/tmp
      - /mnt/GoogleDrive:/mnt/GoogleDrive:ro
      - /usr/local/bin/cue_file:/usr/local/bin/cue_file
    network_mode: "host"
    restart: always
```

1. So I haven't touched any of my folder structure. (that's why it's `/home/radio/liquidsoap/omfm/index.liq`) And it works. But I can't make `hls` work in no way except from `/tmp` directory inside container. And nevertheless it works ok, It seems to me a bit wrong. What do you think? how can I make it work and should I makke it work from another folder (because even `/tmp/younameit` doesn't work). 

2. Also I'm learning about args and Dockerfile now. Any suggestions on that (direct suggestions:).

And I've thought that dockerized LS would take up less disk space. Container is about 360mb or so but it takes up about 1GB:

```ruby
REPOSITORY                TAG       IMAGE ID       CREATED        SIZE
savonet/liquidsoap        v2.2.5    88d40c0b74f6   2 weeks ago    981MB
```

3. Is it normal?

Thank you.

**Reply by vitoyucepi** [reply]
1. > But I can't make hls work in no way except from /tmp directory inside container.

   Check the permissions. The directory should be accessible by the user *liquidsoap:liquidsoap* from inside the container with *7xx* or *x7x* permissions, but this may vary.
1. > Also I'm learning about args and Dockerfile now. Any suggestions on that (direct suggestions:).

   Some time ago I figured out how to build liquidsoap from source in a docker container and described the whole experience in #3609.
1. > 981MB

   Absolute unit, isn't it?
   My ubuntu image from the package - *549 MB*.
   *savonet/liquidsoap-alpine* - *217 MB*.
   My alpine image from source - *181 MB*.
   I can reduce the image size even more by building ffmpeg without modules.

**Reply by gAlleb** [reply]
1. I'll try.
2. Thanks!
3. Absolute unit? Not sure how to understand that. All i did is downloaded it via docker compose up, it showed about 360MB upon download, but upon inspection it takes up 1GB.  https://hub.docker.com/layers/savonet/liquidsoap/v2.2.5/images/sha256-a8d1200b83c5dc92a978d08c2fbdf9bc121729967da0868a2401b25248bebce9?context=explore

Have tried this one savonet/liquidsoap:7f419a2 - just before 2.2.5 release. 
```
savonet/liquidsoap        7f419a2   14b4c8c833bd   2 weeks ago    931MB
```

So, probably they are supposed to be so large if not altered? Or am I missing something?

**Reply by vitoyucepi** [reply]
> Absolute unit

Something very large. I think it's common knowledge that alpine-based images are slimmer than debian-based ones.

> 931MB

2.2.5 has additional packages, I don't know why.

**Reply by gAlleb** [reply]
I see. Well, thank you! What should I do? Gonna learn to build docker image from source?) I'll try)

**Reply by gAlleb** [reply]
Hi @vitoyucepi. Some problems here) So I decided to build `alpine` LS but with ffprobe and python3 so I could use `autocue cue_file`. Everything seems straightforward and looks like this:

```ruby
FROM alpine:edge

CMD ["/bin/sh"]

ARG LIQUIDSOAP_RELEASE="https://github.com/savonet/liquidsoap/releases/download/v2.2.5/liquidsoap-x86_64-2.2.5-r0.apk"
ARG LIQUIDSOAP_SHA256SUM='36437df4ed0ca2046338b509c1100b161f43ab11e8b1427ef97d03269d731459'
RUN wget -O liquidsoap.apk "$LIQUIDSOAP_RELEASE";
RUN echo "$LIQUIDSOAP_SHA256SUM  liquidsoap.apk" | sha256sum -c;
RUN mv liquidsoap.apk /tmp/liquidsoap.apk

USER root

RUN apk add --no-cache ffmpeg
RUN apk add --no-cache python3
RUN apk add --no-cache bash
RUN apk add --allow-untrusted --no-cache \
      -X http://dl-cdn.alpinelinux.org/alpine/edge/testing \
      tini /tmp/liquidsoap.apk && \
      rm -rf /tmp/liquidsoap.apk



USER liquidsoap

ENTRYPOINT ["/sbin/tini" "--" "/usr/bin/liquidsoap"]
```

But LS doesn't work. It's here - https://hub.docker.com/layers/galleb/liquidsoap-alpine-python3-v2.2.5/v0.3/images/sha256-f1c4b09aa38310aa560cc30ec2eba51354139d479823142cfd693285330f86a9?context=repo

```
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/bin/bash: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/bin/bash: not found
```
It's when I added bash to ENTRYPOINT last time. With tini the same:

```
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
```

Moreover If I make image without extra packages so it's identical to official - I get the same results.

Whats wrong? )

**Reply by vitoyucepi** [reply]
I don't know why you would use `alpine:edge` as a base image for a container, but I would prefer ubuntu instead.
Here are parts of my repo. All files are licensed under the GNU AGPLv3.
#### `containerfile`
```containerfile
FROM alpine:3.19 as downloader

ARG LIQUIDSOAP_VERSION=2.2.5

ARG LIQUIDSOAP_RELEASE="https://github.com/savonet/liquidsoap-release-assets/releases/download/v${LIQUIDSOAP_VERSION}"
ARG LIQUIDSOAP_DEB="liquidsoap_${LIQUIDSOAP_VERSION}-ubuntu-jammy-1_amd64.deb"
ARG LIQUIDSOAP_SHA512SUM="56a39e326019ca6f8f25017407f4edac2b1d0b18995e9fba4558e566050c9d48513ab1dbff56034a4ddcc2bed814d51437e400942639605784d2c0e189cc2e0e"

RUN set -eux; \
      mkdir /packages; \
      wget "${LIQUIDSOAP_RELEASE}/${LIQUIDSOAP_DEB}" -O /packages/liquidsoap.deb; \
      echo "${LIQUIDSOAP_SHA512SUM}  /packages/liquidsoap.deb" | sha512sum -c -;

FROM ubuntu:22.04

ENV TZ=UTC

ARG CA_CERTIFICATES_VERSION=20230311ubuntu0.22.04.1
ARG GOSU_VERSION=1.14-1
ARG TZDATA_VERSION=2024a-0ubuntu0.22.04

RUN --mount=type=bind,target=/packages,source=/packages,from=downloader \
    set -eux; \
      apt-get update; \
      apt-get install -y --no-install-recommends \
        ca-certificates="${CA_CERTIFICATES_VERSION}" \
        gosu="${GOSU_VERSION}" \
        tzdata="${TZDATA_VERSION}" \
        /packages/liquidsoap.deb \
      ;

COPY rootfs /

ENTRYPOINT [ "entrypoint" ]
CMD [ "liquidsoap", "/var/liquidsoap/default.liq" ]
```
#### `entrypoint`
```sh
#!/usr/bin/env sh

set -eu

main() {
  if [ "${1#-}" != "$1" ] || [ "${1%.liq}" != "$1" ]; then
    set -- liquidsoap "$@"
  fi

  if [ "$1" = 'liquidsoap' ] && [ "$(id -u)" -eq '0' ]; then
    exec gosu liquidsoap "$0" "$@"
  fi

  exec "$@"
}

main "$@"
```

P.S. @gAlleb, would you mind starting your own personal discussion about containers to keep other discussions tidy? Keep in mind that every participant in the discussion will receive a notification for every new event in the discussion, unless they change their notification settings or explicitly unsubscribe.
